hudi-client/src/test/java/org/apache/hudi/TestHoodieClientOnCopyOnWriteStorage.java

```diff
@@ -515,6 +525,164 @@ public void testSmallInsertHandlingForInserts() throws Exception {
+
+    HoodieWriteClient client = getHoodieWriteClient(config, false);
+
+    // Inserts => will write file1
+    String commitTime1 = "001";
+    client.startCommitWithTime(commitTime1);
+    List<HoodieRecord> inserts1 = dataGen.generateInserts(commitTime1, insertSplitLimit); // this writes ~500kb
+    Set<String> keys1 = HoodieClientTestUtils.getRecordKeys(inserts1);
+    keysSoFar.addAll(keys1);
+    JavaRDD<HoodieRecord> insertRecordsRDD1 = jsc.parallelize(inserts1, 1);
+    List<WriteStatus> statuses = client.upsert(insertRecordsRDD1, commitTime1).collect();
+
+    assertNoWriteErrors(statuses);
+
+    assertEquals("Just 1 file needs to be added.", 1, statuses.size());
+    String file1 = statuses.get(0).getFileId();
+    Assert.assertEquals("file should contain 100 records",
+        readRowKeysFromParquet(jsc.hadoopConfiguration(), new Path(basePath, statuses.get(0).getStat().getPath()))
+            .size(),
+        100);
+
+    // Delete 20 among 100 inserted
+    testDeletes(client, inserts1, 20, file1, "002", 80, keysSoFar);
+
+    // Insert and update 40 records
+    Pair<Set<String>, List<HoodieRecord>> updateBatch2 = testUpdates("003", client, 40, 120);
+    keysSoFar.addAll(updateBatch2.getLeft());
+
+    // Delete 10 records among 40 updated
+    testDeletes(client, updateBatch2.getRight(), 10, file1, "004", 110, keysSoFar);
+
+    // do another batch of updates
+    Pair<Set<String>, List<HoodieRecord>> updateBatch3 = testUpdates("005", client, 40, 150);
+    keysSoFar.addAll(updateBatch3.getLeft());
+
+    // delete non existent keys
+    String commitTime6 = "006";
+    client.startCommitWithTime(commitTime6);
+
+    List<HoodieRecord> dummyInserts3 = dataGen.generateInserts(commitTime6, 20);
+    List<HoodieKey> hoodieKeysToDelete3 = HoodieClientTestUtils
+        .getKeysToDelete(HoodieClientTestUtils.getHoodieKeys(dummyInserts3), 20);
+    JavaRDD<HoodieKey> deleteKeys3 = jsc.parallelize(hoodieKeysToDelete3, 1);
+    statuses = client.delete(deleteKeys3, commitTime6).collect();
+    assertNoWriteErrors(statuses);
+    assertEquals("Just 0 write status for delete.", 0, statuses.size());
+
+    // Check the entire dataset has all records still
+    String[] fullPartitionPaths = new String[dataGen.getPartitionPaths().length];
+    for (int i = 0; i < fullPartitionPaths.length; i++) {
+      fullPartitionPaths[i] = String.format("%s/%s/*", basePath, dataGen.getPartitionPaths()[i]);
+    }
+    assertEquals("Must contain " + 150 + " records", 150,
+        HoodieClientTestUtils.read(jsc, basePath, sqlContext, fs, fullPartitionPaths).count());
+
+    // delete another batch. previous delete commit should have persisted the schema. If not,
+    // this will throw exception
+    testDeletes(client, updateBatch3.getRight(), 10, file1, "007", 140, keysSoFar);
+  }
+
+  private Pair<Set<String>, List<HoodieRecord>> testUpdates(String commitTime, HoodieWriteClient client,
```

Comment:
I would like to leverage this and rewrite the entire test class. But don't want to bloat this diff. Will work on this later. But this would make the tests thinner and makes it easy for readability.

Reply:
Open up a jira for this